keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
+
+## Proposal
+
+#### Risk
+
+The risk with each proposal is:
+
+1. None. Class B) type services which benefitted from being scheduled on nodes
+   with a non-healthy Kube-proxy but which had successfully written the rules
+   required for traffic to reach the endpoint: will not benefit from that
+   anymore. But that's not deterministic...and that's not really a "risk".
+
+2. If the right field indicating "node is terminating / deleting" is
+   chosen/exists: None
+
+#### Mitigations
+
+None, since there are no Risks.
+
+## Design Details
+
+For:
+
+1. Implement change in Kube-proxy which allows the healthcheck server access to
+   the `healthz` state.
+
+2. Possibly implement a Node watcher (I believe it already exists) in
+   Kube-proxy. I also believe there was a discussion about removing said watcher
+   for other reasons, but I am not sure where the consensus lies on that?
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+- Add E2E tests for connectivity to services on terminating nodes and
+  validate graceful termination of TCP connections.
+
+### Graduation Criteria
+
+#### GA
+
+This is addressing a sub-optimal solution currently existing in Kubernetes, and
+the change is deemed safe/harmless enough to implement and turn on by default.
+
+### Upgrade / Downgrade Strategy
+
+No real strategy. The Kube-proxy response between versions might change for
+class B) services, but if it does: it also means that there's a problem which
+the user wasn't aware of before.
+
+### Version Skew Strategy
+
+This doesn't touch load balancer / HC API, so even though an old Kube-proxy
+might talk to a newer control plane, there's no real concern.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+It can't.
+
+###### Does enabling the feature change any default behavior?
+
+Yes, but only for the better. The default behavior might currently indicate to
+load balancers that they should send traffic to endpoints running on nodes,
+where Kube-proxy is not healthy.  
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, by rolling back Kube-proxy.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing. We default back to the old behavior.
+
+###### Are there any tests for feature enablement/disablement?
+
+- Validate that TCP ingress connections are gracefully terminated when
```

Comment:
these tests are going to be tricky to implement

Reply:
Very likely. But we can probably spin up a `agnhost` pod with one eTP:Local service, connect to it through the load balancer IP and call `/shell?cmd=sleep%20infinity` which will just hold the connection to the pod, then taint the node with `ToBeDeletedBy...` and check that our established connection isn't broken. At that time we should also create a new connection and validate that it fails.

Reply:
yeah, the important things is that we do need tests to merge with confidence this really solves the problem statement